Form page - class based admin views

lb- opened the pull request:
Supersedes #3866

Merges in the work from #3866 but also completes functionality and adds testing.

Docs not done yet, I wanted to get this reviewed before adding any doc.s

**Notes**

* Form pages can now declare a custom ListSubmissionsView to override how the submissions list shows (eg. ordering, pagination, csv export filename)
* All ordering handling has been moved to inside the ListSubmissionsView class and changes from PR #3930 has been considered
* All pagination is safe so that incorrect pages in the URL hare handled gracefully (eg. p=9999 will return last page)
* Testing for all of the above
* Credit to go to @frontendr as the bulk of the work has been done by the previous PR

frontendr commented:
@lb- Great work Ben! Thanks for picking this up! I'm currently in a large project so not much time for this. Looks great!

lb- commented:
Checklist

* [ ] linting: wagtail/tests/testapp/models.py:616:1: E302 expected 2 blank lines, found 4
* [ ] re-run CI - getting odd errors: `py36-dj111-postgres-noelasticsearch: commands failed` / `py34-dj111-postgres-noelasticsearch: commands failed`
* [ ] add docs example of custom submissions list view with pagination, ordering and csv filename override

lb- commented:
Issue with the tests I created, when using postgres the ordering of the headings are different for csv export.

**sqllte**

`Username,Submission date,Your email,Chocolate,Ingredients,Your Excitement`

**postgres**

`Username,Submission date,Your Excitement,Ingredients,Chocolate,Your email`

Solution (hopefully) - set field ordering correctly on fields.

lb- commented:
Updated the naming/API a bit - not sure if this is bikeshedding but I think it makes for a more intuitive setup.

loicteixeira commented on wagtail/contrib/forms/views.py:
```diff
@@ -2,152 +2,303 @@
+
+        filtering = self.get_filtering()
+        if filtering:
+            if isinstance(filtering, dict):
```
The two `if`s could be a single condition.

loicteixeira commented on wagtail/contrib/forms/views.py:
```diff
@@ -2,152 +2,303 @@
+    def get_context_data(self, **kwargs):
+        """ Return context for view, handle CSV or normal output """
+        context = super().get_context_data(**kwargs)
+        submissions = context[self.context_object_name].all()  # force queryset into list
```
Why force it here and not leave it to lazily evaluate with the `for` loops below?

lb- commented:
@loicteixeira I have fixed up all those changes except for addressing the concern of loading the list into memory twice.

> However now it does create a list of all the submission data in `get_context_data` and then create the CSV in `get_csv_response` which means the whole submission dataset will be in memory twice.

My understanding of the flow of Django ListView is this:
dispatch -> get -> super get_context_data (which will also call paginate_queryset) -> render_to_response
https://ccbv.co.uk/projects/Django/1.4/django.views.generic.list/ListView/

I was attempting to honour that normal contract for how ListView works, the final split being at the render_to_response method that puts the context through to the csv response. The list of submission data is only ever built once (inside `get_context_data`) and all the `get_csv_response` does is map that to csv rows.

Would it be better to build up the CSV inside the `get_context_data` method?

Happy to take some pointers here, not really sure how to work through this issue.

loicteixeira commented:
Well to be honest, maybe I don't understand how the CSV Writer works under the hood. I though that when you `writerow`, it essentially create the CSV in memory, therefore duplicating the list but maybe it doesn't, it which case it's probably fine.

As for the pagination, I did not realise that the CSV would be paginated as well (but it makes sense since it uses the same list view). Does that mean that if you have 100 submissions, you'll have to export 5 times (since each page has 20 items)? I believe this would change the current behaviour.

The other changes look great but I'll not approve the PR until you can confirm the above question.

lb- commented:
@loicteixeira thanks for the input, you were right about the pagination of CSV. I have now fixed that issue and added a test for CSV to not be paginated.

lb- commented:
@loicteixeira awesome, thanks. should I rebase this in?

Or does anyone else want to take another look?

SalahAdDin commented:
Well done.

lb- commented:
Looks like no issues - I will merge this in soon.

lb- commented:
Merged in https://github.com/wagtail/wagtail/commit/815cb6e405c0671db829279ed418c4a0ed21bbbf

Also closes #3812

Thanks @frontendr & @loicteixeira
